src/3d/symbols/qgspoint3dsymbol_p.cpp

```diff
@@ -0,0 +1,381 @@
+
+  Qt3DRender::QParameter *ambientParameter = new Qt3DRender::QParameter( QStringLiteral( "ka" ), QColor::fromRgbF( 0.05f, 0.05f, 0.05f, 1.0f ) );
+  Qt3DRender::QParameter *diffuseParameter = new Qt3DRender::QParameter( QStringLiteral( "kd" ), QColor::fromRgbF( 0.7f, 0.7f, 0.7f, 1.0f ) );
+  Qt3DRender::QParameter *specularParameter = new Qt3DRender::QParameter( QStringLiteral( "ks" ), QColor::fromRgbF( 0.01f, 0.01f, 0.01f, 1.0f ) );
+  Qt3DRender::QParameter *shininessParameter = new Qt3DRender::QParameter( QStringLiteral( "shininess" ), 150.0f );
+
+  diffuseParameter->setValue( symbol.material().diffuse() );
+  ambientParameter->setValue( symbol.material().ambient() );
+  specularParameter->setValue( symbol.material().specular() );
+  shininessParameter->setValue( symbol.material().shininess() );
+
+  QMatrix4x4 transformMatrix = symbol.transform();
+  QMatrix3x3 normalMatrix = transformMatrix.normalMatrix();  // transponed inverse of 3x3 sub-matrix
+
+  // QMatrix3x3 is not supported for passing to shaders, so we pass QMatrix4x4
+  float *n = normalMatrix.data();
+  QMatrix4x4 normalMatrix4(
+    n[0], n[3], n[6], 0,
+    n[1], n[4], n[7], 0,
+    n[2], n[5], n[8], 0,
+    0, 0, 0, 0 );
+
+  Qt3DRender::QParameter *paramInst = new Qt3DRender::QParameter;
+  paramInst->setName( "inst" );
+  paramInst->setValue( transformMatrix );
+
+  Qt3DRender::QParameter *paramInstNormal = new Qt3DRender::QParameter;
+  paramInstNormal->setName( "instNormal" );
+  paramInstNormal->setValue( normalMatrix4 );
+
+  Qt3DRender::QEffect *effect = new Qt3DRender::QEffect;
+  effect->addTechnique( technique );
+  effect->addParameter( paramInst );
+  effect->addParameter( paramInstNormal );
+
+  effect->addParameter( ambientParameter );
+  effect->addParameter( diffuseParameter );
+  effect->addParameter( specularParameter );
+  effect->addParameter( shininessParameter );
+
+  Qt3DRender::QMaterial *material = new Qt3DRender::QMaterial;
+  material->setEffect( effect );
+
+  return material;
+}
+
+void QgsPoint3DSymbolInstancedEntityFactory::addEntityForSelectedPoints( const Qgs3DMapSettings &map, QgsVectorLayer *layer, const QgsPoint3DSymbol &symbol, QgsPoint3DSymbolEntity *parent )
+{
+  // build the default material
+  Qt3DRender::QMaterial *mat = material( symbol );
+
+  // update the material with selection colors
+  Q_FOREACH ( Qt3DRender::QParameter *param, mat->effect()->parameters() )
+  {
+    if ( param->name() == "kd" ) // diffuse
+      param->setValue( map.selectionColor() );
+    else if ( param->name() == "ka" ) // ambient
+      param->setValue( map.selectionColor().darker() );
+  }
+
+  // build the feature request to select features
+  QgsFeatureRequest req;
+  req.setDestinationCrs( map.crs() );
+  req.setFilterFids( layer->selectedFeatureIds() );
+
+  // build the entity
+  QgsPoint3DSymbolInstancedEntityNode *entity = new QgsPoint3DSymbolInstancedEntityNode( map, layer, symbol, req );
```

Comment:
bit confused - what happens to ownership of entity?

Reply:
It gets parented to another entity two lines below...